File: node/src/symmetricstate.ts
import { hash } from '@stablelib/sha256';
import { concat } from 'uint8arrays/concat';
import { CipherAlgorithm, Bytes, hkdf } from './crypto';
import { CipherState } from './cipherstate';

const EMPTY = new Uint8Array(0);

export class SymmetricState {
    public ck: Bytes = EMPTY;
    public h: Bytes = EMPTY;
    public cs: CipherState | null = null;

    public mixHash(data: Bytes) {
        this.h = hash(concat([this.h, data]));
    }

    public mixKey(cipher: CipherAlgorithm, key: Bytes) {
        const [ck, temp] = hkdf(key, this.ck);
        this.ck = ck;
        this.cs = new CipherState(cipher, temp);
    }

    public async encryptWithAd(plaintext: Bytes, ad: Bytes): Promise<Bytes> {
        const cs = this.cs!;
        const ciphertext = await cs.cipher.seal(cs.key, cs.nonce.getBytes(), plaintext, ad);
        cs.nonce.increment();
        return ciphertext;
    }

    public async encryptAndMixHash(plaintext: Bytes, mustSecret: boolean): Promise<Bytes> {
        let ciphertext: Uint8Array;
        // if key mixed
        if (this.cs) {
            ciphertext = await this.encryptWithAd(plaintext, this.h);
        } else {
            if (mustSecret) {
                throw new Error('must secret');
            }
            ciphertext = plaintext;
        }
        this.mixHash(ciphertext);
        return ciphertext;
    }

    // throwable
    public async decryptWithAd(ciphertext: Bytes, ad: Bytes): Promise<Bytes> {
        const cs = this.cs!;
        const plaintext = await this.cs!.cipher.open(cs.key, cs.nonce.getBytes(), ciphertext, ad);
        cs.nonce.increment();
        return plaintext;
    }

    // throwable
    public async mixHashAndDecrypt(ciphertext: Bytes): Promise<Bytes> {
        if (this.cs) {
            const plaintext = this.decryptWithAd(ciphertext, this.h);
            this.mixHash(ciphertext);
            return plaintext;
        } else {
            this.mixHash(ciphertext);
            return ciphertext;
        }
    }
}

File: java/src/main/kotlin/kr/jclab/jscp/state/SymmetricState.kt
package kr.jclab.jscp.state

import kr.jclab.jscp.crypto.CipherAlgorithm
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import java.security.MessageDigest

open class SymmetricState {
    private var ck: ByteArray = ByteArray(0)
    private var h: ByteArray = ByteArray(0)
    private var cs: CipherState? = null

    open fun mixHash(data: ByteArray) {
        val digest = MessageDigest.getInstance("SHA-256")
        h = digest.digest(h + data)
    }

    open fun mixKey(cipher: CipherAlgorithm, key: ByteArray) {
        val (newCk, temp) = hkdf(key, ck)
        ck = newCk
        cs = CipherState(cipher, temp)
    }

    fun encryptWithAd(plaintext: ByteArray, ad: ByteArray): ByteArray {
        val state = cs ?: throw IllegalStateException("No cipher state")
        val ciphertext = state.cipher.seal(state.key, state.nonce.bytes, plaintext, ad)
        state.nonce.increment()
        return ciphertext
    }

    fun encryptAndMixHash(plaintext: ByteArray, mustSecret: Boolean): ByteArray {
        val ciphertext = if (cs != null) {
            encryptWithAd(plaintext, h)
        } else {
            if (mustSecret) throw IllegalStateException("Must be secret")
            plaintext
        }
        mixHash(ciphertext)
        return ciphertext
    }

    fun decryptWithAd(ciphertext: ByteArray, ad: ByteArray): ByteArray {
        val state = cs ?: throw IllegalStateException("No cipher state")
        val plaintext = state.cipher.open(state.key, state.nonce.bytes, ciphertext, ad)
        state.nonce.increment()
        return plaintext
    }

    fun mixHashAndDecrypt(ciphertext: ByteArray): ByteArray {
        val plaintext = if (cs != null) {
            decryptWithAd(ciphertext, h)
        } else {
            ciphertext
        }
        mixHash(ciphertext)
        return plaintext
    }

    private fun hkdf(ikm: ByteArray, salt: ByteArray): Pair<ByteArray, ByteArray> {
        val hkdf = HKDFBytesGenerator(SHA256Digest())
        val params = HKDFParameters(ikm, salt, null)
        hkdf.init(params)

        val okm = ByteArray(96)
        hkdf.generateBytes(okm, 0, okm.size)

        return Pair(
            okm.copyOfRange(0, 32),
            okm.copyOfRange(32, 64)
        )
    }
}